### ISSUE TITLE
Compile failure on older OS X clang with OpenSSL 1.0.2e

### ISSUE THREAD

**Issue by reaperhulk** [openssl/openssl#494]
https://git.openssl.org/gitweb/?p=openssl.git;a=commitdiff;h=301a6dcd4590fb2f69d08259577e215b4cc3caa3#patch5 added a check to see if it should use the ADDX instructions based on the clang version. Unfortunately, on older versions of clang on OS X this check incorrectly returns true and compilation then fails with errors like:

```
x86_64-mont.s:966:2 error: invalid instruction mnemonic 'adcxq'
adcxq %rax,%r12
^~~~~
```
The version of the compiler in question is: `Apple LLVM version 4.2 (clang-425.0.28) (based on LLVM 3.2svn)` (On an ancient OS X 10.7 VM)

**Comment by mvkorpel** [comment]
I also bumped into this issue when upgrading the Homebrew installation on my MacBook (early 2008) running OS X 10.7.5. The compiler version is the same as above, and that is the last one provided by Apple for this OS version. I was able to compile the 64-bit library by undoing the changes of the commit in question.

**Comment by mojca** [comment]
This seems to be caused by https://github.com/openssl/openssl/commit/76eba0d94bb418325be6409b272eac5e2bd4a0a9 after [this bug report](https://mta.openssl.org/pipermail/openssl-bugs-mod/2015-November/000248.html). You can check the clang reported version strings [here](http://trac.macports.org/wiki/XcodeVersionInfo).

The latest version of clang shipped by Apple that can be used on 10.7:
```
Apple LLVM version 4.2 (clang-425.0.28) (based on LLVM 3.2svn)
```

Recent versions of clang shipped by Apple report just:
```
Apple LLVM version 7.0.0 (clang-700.1.76)
```
where the string "based on LLVM" has been removed.

The string returned by self-compiled clang (version 3.4):
```
clang version 3.4.2 (tags/RELEASE_34/dot2-final)
```

Now the file `crypto/bn/asm/x86_64-mont.pl` checks whether version of clang is greater than 3.3. Clang on 10.7 is `Apple LLVM version 4.2 (clang-425.0.28) (based on LLVM 3.2svn)`. The previous regexp
```
/(^clang version|based on LLVM) ([3-9])\.([0-9]+)/
```
picked `based on LLVM 3.2`, but the new regexp
```
/((?:^clang|LLVM) version|based on LLVM) ([3-9])\.([0-9]+)/
```
picks up
```
LLVM version 4.2
```
where 4.2 is greater than 3.3, but wrong and results in code that is not supported by the compiler.

**Comment by mvkorpel** [comment]
Something like the following regular expressions might fix this. The idea is to use the last suitable version number appearing in the compiler version string. I removed the minimum major version requirement (2 or 3) from the regex so that the logic would be correct in cases where a version number in the string meets the requirement but the last one does not. The extra matches that this causes should not matter because the version number is checked afterwards anyway.

Major and minor versions captured together:
`(?:((?:^clang|.*LLVM) version|.*based on LLVM) ([0-9]\.[0-9]+)[^ ]*)+`

Major and minor versions captured separately:
`(?:((?:^clang|.*LLVM) version|.*based on LLVM) ([0-9])\.([0-9]+)[^ ]*)+`

**Comment by mojca** [comment]
I don't know. I have a feeling that this is prone to fail at some other point in the future. The code really needs to distinguish between the "arbitrary" `Apple clang version 4.1/Apple LLVM version 4.2/.../LLVM version 7.0.0` (based on Xcode version) and `based on LLVM 3.[x]svn` which is the true clang version – usually the same as `clang version 3.[x]` that any other clang compiler would report. If the source code wants to enable some feature that has only been added in some (future) version of clang 4.[y], then `Apple LLVM version 7.0.0` would be a false positive even with your proposal. It's perfectly OK if the developers decide to base their decision on the version of Xcode, but this then has to be at least consistent.

I suspect that this requires a slightly more complex solution.

**Comment by mvkorpel** [comment]
I agree that my proposed solution may not be robust for the future, but as far as I see it should work as a quick drop-in replacement for the short term. One case where the regular expressions will fail (in the future) is if the major version (be it Xcode or clang originated) is 10 or above. That would result in the compiler being detected as less capable than it really is, which would "only" affect runtime performance but not make the build fail. I wonder why the major version number is only allowed to have one digit.
> If the source code wants to enable some feature that has only been added in some (future) version of clang 4.[y], then `Apple LLVM version 7.0.0` would be a false positive even with your proposal.

True, but speculative.

**Comment by neverpanic** [comment]
https://trac.macports.org/attachment/ticket/47035/fix-Apple-clang-version-detection.patch might fix this without mixing the version number space between "Apple clang" and clang. I'll wait for feedback from users, though, because I cannot test this on 10.7.

**Comment by boogaweb** [comment]
How can I test it, I am on 10.7.5 and had this issue with openssl installing tor. I am not really into GIT processes and techniques, BTW

**Comment by neverpanic** [comment]
@boogaweb You need to apply this patch to the OpenSSL source tree before compiling it.

Users at MacPorts confirm the patch fixes the problem for them, so I have applied it downstream now. I have added a header to the patch file. You can always grab the latest version at https://svn.macports.org/repository/macports/trunk/dports/devel/openssl/files/fix-Apple-clang-version-detection.patch.

Should I open a pull request for this?

**Comment by boogaweb** [comment]
As I am not very skilled on patching a macport app (never did it before), i will have to wait for the updated app... :(

**Comment by dot-asm** [comment]
It was addressed in e780ed033659cb1269a49c8952517559cbd76558 by giving "based on LLVM" higher priority.

**Comment by mojca** [comment]
Just curious: why don't you put this version checking to a single place and perhaps call that function rather than requiring a change (and code repetition) at 13 different places?
